neurovault/urls.py

```diff
@@ -380,6 +389,7 @@ class CollectionViewSet(mixins.RetrieveModelMixin,
     filter_fields = ('name', 'DOI', 'owner')
     serializer_class = CollectionSerializer
     filter_backends = (DjangoFilterBackend,)
+    permission_classes = (permissions.IsAuthenticatedOrReadOnly,)
```

Comment:
Yeah, I tried that. Problem is, when we're handling image upload in `CollectionViewSet`, DRF first checks permissions for the `CollectionViewSet` while [dispatching a request](https://github.com/tomchristie/django-rest-framework/blob/4e5da16961c8f11d01a2f9dc475efacb29e778c7/rest_framework/views.py#L454) and before calling the handler. So when it hits `APIView.check_permissions()`, it calls [`has_permission`](https://github.com/tomchristie/django-rest-framework/blob/4e5da16961c8f11d01a2f9dc475efacb29e778c7/rest_framework/views.py#L318) (and not [`has_object_permission`](https://github.com/tomchristie/django-rest-framework/blob/4e5da16961c8f11d01a2f9dc475efacb29e778c7/rest_framework/views.py#L329)). `DjangoObjectPermissions` doesn't implement `has_permission()` but since it's derived from `DjangoModelPermissions`, `DjangoModelPermissions.has_permissions()` gets called. And it's irrelevant to the job.

[As DRF's documentation says](http://www.django-rest-framework.org/api-guide/permissions/#custom-permissions): "The instance-level `has_object_permission` method will only be called if the view-level `has_permission` checks have already passed." The `permission_classes` in question is a view-level attribute.

Still even if we disable the irrelevant view-level checks, I can't see any way to trigger `has_object_permission` in the `detail_route` other than calling [`GenericAPIView.get_object`](https://github.com/tomchristie/django-rest-framework/blob/6284bceaaff0e53349131164ce5c16cda8deb715/rest_framework/generics.py#L76) (instead of `get_collection`.)

So I decided to go with an explicit check and used the same method for exception raising as in `APIView.check_object_permissions `. 

While writing this I realized that there's another option available (for more implicit checks):
1. Set `permission_classes` to `DjangoObjectPermissions`, as you suggest.
2. In the `CollectionViewSet.add_image` endpoint, call `GenericAPIView.get_object()` to fetch the collection object (this will trigger `check_object_permissions`.)
3. Use PUT method instead of POST for images, because it maps to `change_collection` in [`DjangoObjectPermissions.perms_map`](https://github.com/tomchristie/django-rest-framework/blob/6284bceaaff0e53349131164ce5c16cda8deb715/rest_framework/permissions.py#L88).
4. Use Django's model permissions system to pass the view-level `has_permission` check. (I can't find any trace of a model permissions usage in the code. But I may miss something.)